engine/servermaster/config.go

```diff
@@ -60,8 +60,9 @@ const (
 type Config struct {
 	LogConf logutil.Config `toml:"log" json:"log"`
 
-	Addr          string `toml:"addr" json:"addr"`
-	AdvertiseAddr string `toml:"advertise-addr" json:"advertise-addr"`
+	Addr                string `toml:"addr" json:"addr"`
+	AdvertiseAddr       string `toml:"advertise-addr" json:"advertise-addr"`
+	AdvertiseLeaderAddr string `toml:"advertise-leader-addr" json:"advertise-leader-addr"`
```

Comment:
Could you please explain why we separate these two addresses. In K8s environment, maybe neither `AdvertiseAddr` nor `AdvertiseLeaderAddr` is exposed to end user directly.

Reply:
`advertise-leader-addr` is like etcd's `advertise-peer-addr`. It is used for followers forwarding requests to the leader.  In K8s environment, k8s operator will create a service for servermaster, all clients should use this address, instead of the pod address (`advertise-leader-addr`). Otherwise, clients can't obtain the benefits of k8s service load balancers.

Reply:
On the other hand, the service endpoint is usually fixed. Separating these two addresses can simplify the deployment of executors.

Reply:
It's a command line parameter and we cannot change it once the process started. But the leadership of the cluster will be changed dynamically. How do we guarantee the address is correct all the time？

Reply:
We can reference the pod ip or the pod name. The pod has the same lifetime as servermaster process. So the address is always correct.

Reply:
Does it mean in K8s environment, given the following topology 
- A service of server master, such as `svc-server-master:10000`
- Three server master pods
   - Each of them binds `advertise-addr` to `svc-server-master:10000`
   - Each of them binds `advertise-leader-addr` to a headless service, such as `svc-server-master-<id>:10001`
- All executors communicate with `svc-server-master:10000`

Reply:
> Does it mean in K8s environment, we can deploy as follows:
> 
> * A service, naming `svc-server-master:10000`, with selector of server master
> * Three server master pods
>   
>   * Each of them binds `advertise-addr` to `svc-server-master:10000`
>   * Each of them binds `advertise-leader-addr` to a headless service, such as `svc-server-master-<id>:10000`
> * All executor pods communicate with `svc-server-master:10000`

@amyangfei Yes, that's right.